### PR TITLE
fix(text): refact text.dm for Unicode support

### PR DESCRIPTION
**Pull request by Epicus7**
Transfer text.dm changes from russian Baystation fork "OnyxBay" for Unicode support.

In our community we have encountered many problems with Russian language and other unicode characters (hooray, emojis! 🙃), while trying to host Russian Aurora server. Here our fixes according to [BYOND's Unicode guidelines](http://www.byond.com/docs/ref/#/{notes}/Unicode).

`/proc/sanitizeName` proc has mostly been ported from /tg/station13.

### CODE REVIEW

**Comment by Epicus7** [comment]
!review

File: code/_helpers/text.dm [hunk 1]
```diff
@@ -57,76 +57,116 @@
...
-	var/number_of_alphanumeric	= 0
-	var/last_char_group			= 0
+/**
+  * Filters out undesirable characters from names.
+  *
+  * * strict - return null immidiately instead of filtering out  TODO: port from tg
+  * * allow_numbers - allows numbers and common special characters - used for silicon/other weird things names
+  */
+
+/proc/sanitizeName(input, max_length = MAX_NAME_LEN, allow_numbers = 0, force_first_letter_uppercase = TRUE)
```
Comment (PlinHost) [inline code/_helpers/text.dm:73]: `allow_numbers` default was changed from 1 to 0, new `force_first_letter_uppercase` parameter is not provided anywhere nor used in this proc.

File: code/_helpers/text.dm [hunk 2]
```diff
@@ -57,76 +57,116 @@
...
+		return //protects against tiny names like "A" and also names like "' ' ' ' ' ' ' '"
 
-	if(last_char_group == 1)
-		output = copytext(output,1,length(output))	//removes the last character (in this case a space)
+	if(last_char_group == SPACES_DETECTED)
+		output = copytext_char(output, 1, -1) //removes the last character (in this case a space)
 
-	for(var/bad_name in list("space","floor","wall","r-wall","monkey","unknown","inactive ai","plating"))	//prevents these common metagamey names
-		if(cmptext(output,bad_name))	return	//(not case sensitive)
+	for(var/bad_name in list("space","floor","wall","r-wall","monkey","unknown","inactive ai"))	//prevents these common metagamey names
```
Comment (PlinHost) [inline code/_helpers/text.dm:152]: "plating" was removed from bad name list.

File: code/_helpers/text.dm (hunk 1 above)
Comment (Epicus7) [reply code/_helpers/text.dm]: Changed default value, removed `force_first_letter_uppercase`

File: code/_helpers/text.dm (hunk 2 above)
Comment (Epicus7) [reply code/_helpers/text.dm]: Returned `plating` back

**Pull request merged**